crates/miden-lib/src/note/well_known_note.rs

```diff
@@ -145,4 +147,84 @@ impl WellKnownNote {
             },
         }
     }
+
+    /// Checks the correctness of the provided note inputs against the target account.
+    ///
+    /// It performs:
+    /// - for all notes: a check that note inputs have correct number of values.
+    /// - for `P2ID` note: assertion that the account ID provided by the note inputs is equal to the
+    ///   target account ID.
+    /// - for `P2IDR` note: assertion that the account ID provided by the note inputs is equal to
+    ///   the target or sender account IDs.
+    pub fn check_note_inputs(
+        &self,
+        note: &Note,
+        target_account_id: AccountId,
+        block_ref: BlockNumber,
+    ) -> NoteAccountCompatibility {
+        match self {
+            WellKnownNote::P2ID => {
+                let note_inputs = note.inputs().values();
+                if note_inputs.len() != 2 {
+                    return NoteAccountCompatibility::No;
+                }
+
+                Self::check_input_account_id(note_inputs, &[target_account_id])
+            },
+            WellKnownNote::P2IDR => {
+                let note_inputs = note.inputs().values();
+                if note_inputs.len() != 3 {
+                    return NoteAccountCompatibility::No;
+                }
+
+                let recall_height: Result<u32, _> = note_inputs[2].try_into();
+                let Ok(recall_height) = recall_height else {
+                    return NoteAccountCompatibility::No;
+                };
+
+                if block_ref.as_u32() >= recall_height {
+                    let sender_account_id = note.metadata().sender();
+                    // if the sender can already reclaim the assets back, provide it as a
+                    // valid account alongside with the target account
+                    Self::check_input_account_id(
+                        note_inputs,
+                        &[target_account_id, sender_account_id],
+                    )
```

Comment:
Is this actually correct? My understanding is that this would check that the account ID encoded in `note_inputs` is either `target_account_id` or `sender_account_id` - but I don't think that this is what we actually want. I think what we want is to check that:

- Either `target_account_id == inputs_account_id`.
- Or `target_account_id == sender_account_id`.

Reply:
Ah, I see, I confused which ID is changing. Right, so if reclaim hight is not reached yet the note should contain the target ID, and this ID should also be a transaction target. But if the reclaim hight was reached the account against which the transaction is executed could be the same as sender account, but anyway the note remains the same. I'll update this check to make it correct.